bigquery/google/cloud/bigquery/table.py

```diff
@@ -783,43 +784,72 @@ def __init__(self, values, field_to_index):
         self._xxx_field_to_index = field_to_index
 
     def values(self):
-        return self._xxx_values
+        """Return the values included in this row.
 
-    def keys(self):
+        Returns:
+            Sequence[object]: A sequence of length ``len(row)``.
         """
-        Return keys as of a dict:
-        >>> Row(('a', 'b'), {'x': 0, 'y': 1}).keys()
-        ['x', 'y']
+        return copy.deepcopy(self._xxx_values)
+
+    def keys(self):
+        """Return the keys for using a row as a dict.
+
+        Returns:
+            Sequence[str]: The keys corresponding to the columns of a row
+
+        Examples:
+
+            >>> list(Row(('a', 'b'), {'x': 0, 'y': 1}).keys())
+            ['x', 'y']
         """
-        keys = self._xxx_field_to_index.keys()
-        return keys
+        return six.iterkeys(self._xxx_field_to_index)
 
     def items(self):
+        """Return items as ``(key, value)`` pairs.
+
+        Returns:
+            Sequence[Tuple[str, object]]:
+                The ``(key, value)`` pairs representing this row.
+
+        Examples:
+
+            >>> list(Row(('a', 'b'), {'x': 0, 'y': 1}).items())
+            [('x', 'a'), ('y', 'b')]
         """
-        Return items as of a dict:
-        >>> Row(('a', 'b'), {'x': 0, 'y': 1}).items()
-        [('x', 'a'), ('y', 'b')]
-        """
-        items = [
-            (k, self._xxx_values[i])
-            for k, i
-            in self._xxx_field_to_index.items()
-        ]
-        return items
+        for key, index in six.iteritems(self._xxx_field_to_index):
+            yield (key, copy.deepcopy(self._xxx_values[index]))
```

Comment:
Here the copy might make more sense, because even if `self._xxx_values` is a `tuple` (i.e. immutable), it's values might not be.